capsules/src/analog_comparator.rs

```diff
@@ -0,0 +1,156 @@
+        }
+        // Convert channel index
+        let chan = self.channels[channel];
+        let result = self.analog_comparator.comparison(chan);
+
+        return ReturnCode::SuccessWithValue {
+            value: result as usize,
+        };
+    }
+
+    // Start comparing on a channel
+    fn start_comparing(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
+        }
+        // Convert channel index
+        let chan = self.channels[channel];
+        let result = self.analog_comparator.start_comparing(chan);
+
+        return result;
+    }
+
+    // Stop comparing on a channel
+    fn stop_comparing(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
+        }
+        // Convert channel index
+        let chan = self.channels[channel];
+        let result = self.analog_comparator.stop_comparing(chan);
+
+        return result;
+    }
+}
+
+impl<'a, A: hil::analog_comparator::AnalogComparator> Driver for AnalogComparator<'a, A> {
+    /// Control the analog comparator.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Perform a simple comparison.
+    ///        Input x chooses the desired comparator ACx (e.g. 0 or 1 for
+    ///        hail, 0-3 for imix)
+    /// - `2`: Start interrupt-based comparisons.
+    ///        Input x chooses the desired comparator ACx (e.g. 0 or 1 for
+    ///        hail, 0-3 for imix)
+    /// - `3`: Stop interrupt-based comparisons.
+    ///        Input x chooses the desired comparator ACx (e.g. 0 or 1 for
+    ///        hail, 0-3 for imix)
+    fn command(&self, command_num: usize, channel: usize, _: usize, _: AppId) -> ReturnCode {
+        match command_num {
+            0 => ReturnCode::SuccessWithValue {
+                value: self.channels.len() as usize,
+            },
+
+            1 => self.comparison(channel),
+
+            2 => self.start_comparing(channel),
+
+            3 => self.stop_comparing(channel),
+
+            _ => return ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        _app_id: AppId,
+    ) -> ReturnCode {
+        match subscribe_num {
+            // Subscribe to all interrupts
+            0 => {
+                self.callback.set(callback);
+                ReturnCode::SUCCESS
+            }
+            // Default
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+}
+
+impl<'a, A: hil::analog_comparator::AnalogComparator> hil::analog_comparator::Client
+    for AnalogComparator<'a, A>
+{
+    fn fired(&self, channel: usize) {
+        // Callback to userland
+        self.callback
+            .get()
+            .map_or_else(|| false, |mut cb| cb.schedule(channel, 0, 0));
```

Comment:
If the array of channels passed to `new()` is re-ordered, does this still work? I'm not sure how the index is associated with the channel.